design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,314 @@
+@UdfReturn(value  = "STRUCT<'A' INT, 'B' VARCHAR>") // sample specification annotation
+@UdfReturn(file   = 'schema_def.kschema')           // another way pointing to a file
+@UdfReturn(schema = "name.space.MyClass.SCHEMA")    // another way that would resolve a java object
+public Struct generate() {
+  return new Struct(...);
+}
+```
+
+### Generics
+
+Today, we eagerly convert UDF output types to corresponding Connect schema types at the time that we
+compile the UDFs and not the time that we use them in a `SELECT` clause. Since there is no 
+corresponding "unresolved" type in Connect, we fail this conversion. To solve this problem, we can
+infer the type from the source schema and compile the code to work with any component type (e.g.
+omit the generics from the compiled code).
+
+```java
+@Udf("returns a sublist of 'list'")
+public <T> List<T> sublist(
+    @UdfParameter final List<T> list, 
+    @UdfParameter final int start, 
+    @UdfParameter final int end) {
+  return list.subList(start, end);
+}
+```
+
+We need to ensure that the output type can be inferred. This means that we need to fail if the 
+signature is something like `<T> List<T> convert(List<String> list)`. This can be done during 
+compilation by following a simple rule: any generic type used in the output must be present in at
+least one of the parameters. We can access this information via reflection.
+
+**NOTE:** Supporting a wildcard output type is not covered by this design since we would not be able 
+to generate the output schema for select statements, however supporting wildcards in the input types 
+(e.g. `Long length(List<?> list)`) should be possible.
+
+### Varargs
+
+Varargs boils down to supporting native Java arrays as parameters and ensuring that component types 
+are resolved properly. Anytime a method is registered with variable args (e.g. `String... args`) we 
+will register a corresponding function with the wrapping array type (e.g. `String[]`). At runtime, 
+we resolve methods using the following fallback logic:
+```
+Type[] desired
+if signature_exists(desired): 
+  return method
+else:
+  vararg = desired[-1]
+  while desired[-1] == vararg:
+    if signature_exists(desired, arrayOf(vararg)) return method 
+    desired = desired[:-1]
+  fail
+```
+This will allows us to resolve methods such as `foo(Int, String...)` and `foo(String, String...)`,
+as well as accept empty arguments to `foo(String...)`. If any parameter is `null`, it will be 
+considered valid for any vararg declaration.
+
+This proposal does not cover supporting `Object` as a parameter to a method in order to allow for 
+"generic variable argument" UDFs, but supporting it can be an extension.
+
+```java
+@Udf("returns a sublist of 'list' t")
+public int sum(@UdfParameter final int... args) {
+  return Arrays.stream(args).sum();
+}
+```
+
+### Complex Aggregation
+
+We will allow users to supply an additional method `VR export(A agg)`. This method will be taken
```

Comment:
I'm not understanding why this is necessary. Why can't we just change our aggregations to support arbitrary numbers of arguments, and arbitrary return types. Why do we need this extra export step. The fact that we only support 1 parameter which requires the same type is because we've hard-coded it this way. May be easier to discuss in person.

Reply:
You might not want to aggregate on the value that you return. For example, in the `avg` case you want to aggregate on `struct{long, long}` but want to return `double`. Let's talk in person